packages/teleport/src/Discover/DownloadScript/useDownloadScript.tsx

```diff
@@ -22,20 +22,120 @@ import { AgentStepProps } from '../types';
 
   useEffect(() => {
+    if (!retry && pollState === 'polling') return;
+
+    // Set flags to default.
+    setRetry(false);
+    setPollState('polling');
+
+    // abortController is required to cancel any in flight request.
+    const abortController = new AbortController();
+    const abortSignal = abortController.signal;
+    let timeoutId;
+    let intervalId;
+
+    // inFlightReq is a flag to prevent another fetch request when a
+    // previous fetch request is still in progress. May happen when a
+    // previous fetch request is taking longer than the polling
+    // interval time.
+    let inFlightReq;
+
+    function fetchNodeMatchingRefResourceId(token: JoinToken) {
+      if (inFlightReq) return;
+
+      inFlightReq = ctx.nodesService
+        .fetchNodes(
+          ctx.clusterId,
+          {
+            search: `${token.refResourceId}`,
+            limit: 1,
+          },
+          abortSignal
+        )
+        .then(res => {
+          if (res.agents.length > 0) {
+            setPollState('success');
+            props.updateAgentMeta({
+              ...props.agentMeta,
+              refResourceId: token.refResourceId,
+              node: res.agents[0],
+            });
+            cleanUp();
+          }
+        })
+        // Polling related errors are ignored.
+        // The most likely cause of error would be network issues
+        // and aborting in flight request.
+        .catch(() => {})
+        .finally(() => {
+          inFlightReq = null; // reset flag
+        });
+    }
+
+    function cleanUp() {
+      clearInterval(intervalId);
+      clearTimeout(timeoutId);
+      // Cancel any in flight request.
+      abortController.abort();
+    }
+
     run(() =>
-      ctx.joinTokenService.fetchJoinToken(['Node'], 'token').then(setJoinToken)
+      ctx.joinTokenService.fetchJoinToken(['Node'], 'token').then(token => {
+        // Probably will never happen, but just in case, otherwise
+        // querying for the resource can return a false positive.
+        if (!token.refResourceId) {
+          setAttempt({
+            status: 'failed',
+            statusText:
+              'reference resource ID is required to discover the newly added resource',
+          });
+          return;
+        }
+
+        setJoinToken(token);
+
+        // Start the poller to discover the resource just added.
+        intervalId = setInterval(
+          () => fetchNodeMatchingRefResourceId(token),
+          THREE_SECONDS_IN_MS
+        );
+
+        // Set a timeout in case polling continuosly produces
+        // no results. Which means there is either a network error,
+        // script is ran unsuccessfully, script has not been ran,
+        // or resource cannot connect to cluster.
+        timeoutId = setTimeout(() => {
+          setPollState('error');
+          cleanUp();
+        }, FIVE_MINUTES_IN_MS);
+      })
     );
-  }, []);
+
+    return () => {
+      cleanUp();
+    };
+  }, [retry]);
```

Comment:
Are we ok with skipping hooks dependencies? Eg: currently we will only rerun this effect when `retry` boolean changes. We are ignoring: `pollState`, `ctx.nodesService.fetchNodes`, `ctx.clusterId`, `props.updateAgentMeta`, `props.agentMeta` etc.

Reply:
We already have that package installed, looks like the correct config isn't enabled though, thx for the reminder.

Reply:
so about this, i know it's an eslint rule, but from prior discussions I had with @alex-kovoy-dd, this rule is just a safeguard and a hint that there might be a bug. As long as we are intentional and know what we are doing, it was okay to ignore this rule. I also feel like providing an exhaustive dependency list is more confusing, but if folks are against this I can start providing the exhaustive list.

Reply:
I agree with @JanKaczmarkiewicz it would better to not skip the dependencies (and enable the plugin in the future). I don't see much benefit from skipping them, as we risk potential bugs that may not be visible at first glance. React docs also discourage this. In addition, adding all dependencies reduces the number of decisions because we do not have to think about whether a particular dependency should be added or not.
I'd try to provide exhaustive lists of dependencies, so it will be easier to enable the plugin.
Nice art: [Don’t Lie to React About Dependencies](https://overreacted.io/a-complete-guide-to-useeffect/#dont-lie-to-react-about-dependencies)